File: apps/web-enterprise/src/lib/api/calendar.ts

import { api } from '@/lib/api/config'
import { Task, TaskStatus, TaskPriority, TaskType } from "./tasks"

// ============================================================================
// Calendar Event Types (matching backend)
// ============================================================================

export interface CalendarEvent {
  id: string
  title: string
  description?: string
  event_type: EventType
  status: EventStatus
  start_date: string  // ISO 8601
  end_date: string    // ISO 8601
  is_all_day: boolean
  timezone: string
  location?: string
  meeting_url?: string
  meeting_id?: string
  meeting_password?: string
  creator_id?: string
  task_id?: string
  project_id?: string
  okr_id?: string
  recurrence_type: RecurrenceType
  recurrence_pattern?: RecurrencePattern
  recurrence_end_date?: string
  parent_event_id?: string
  is_private: boolean
  is_important: boolean
  reminder_minutes?: number
  color: string
  external_id?: string
  external_source?: string
  meta_data?: Record<string, any>
  org_id: string
  created_at: string
  updated_at: string
  deleted_at?: string
  is_deleted: boolean
  attendees?: EventAttendee[]
  duration_minutes?: number
}

export interface EventAttendee {
  id: string
  event_id: string
  user_id?: string
  email?: string
  display_name?: string
  status: AttendeeStatus
  is_organizer: boolean
  is_required: boolean
  response_date?: string
  notes?: string
  created_at: string
  updated_at: string
}

export interface RecurrencePattern {
  interval: number
  days_of_week?: number[]  // 0=Monday, 6=Sunday
  day_of_month?: number
  month_of_year?: number
  count?: number
  exceptions?: string[]  // ISO dates to exclude
}

export enum EventType {
  MEETING = "meeting",
  TASK_DEADLINE = "task_deadline",
  PROJECT_MILESTONE = "project_milestone",
  OKR_REVIEW = "okr_review",
  PERSONAL = "personal",
  HOLIDAY = "holiday",
  VACATION = "vacation",
  OTHER = "other"
}

export enum EventStatus {
  CONFIRMED = "confirmed",
  TENTATIVE = "tentative",
  CANCELLED = "cancelled",
  COMPLETED = "completed"
}

export enum RecurrenceType {
  NONE = "none",
  DAILY = "daily",
  WEEKLY = "weekly",
  MONTHLY = "monthly",
  YEARLY = "yearly",
  CUSTOM = "custom"
}

export enum AttendeeStatus {
  PENDING = "pending",
  ACCEPTED = "accepted",
  DECLINED = "declined",
  TENTATIVE = "tentative"
}

// ============================================================================
// API Request/Response Types
// ============================================================================

export interface CreateEventRequest {
  title: string
  description?: string
  event_type: EventType
  start_date: string  // ISO 8601
  end_date: string    // ISO 8601
  is_all_day?: boolean
  timezone?: string
  location?: string
  meeting_url?: string
  meeting_id?: string
  meeting_password?: string
  task_id?: string
  project_id?: string
  okr_id?: string
  recurrence_type?: RecurrenceType
  recurrence_pattern?: RecurrencePattern
  recurrence_end_date?: string
  is_private?: boolean
  is_important?: boolean
  reminder_minutes?: number
  color?: string
  meta_data?: Record<string, any>
  attendees?: CreateEventAttendee[]
}

export interface CreateEventAttendee {
  user_id?: string
  email?: string
  display_name?: string
  is_required?: boolean
  notes?: string
}

export interface UpdateEventRequest {
  title?: string
  description?: string
  event_type?: EventType
  status?: EventStatus
  start_date?: string
  end_date?: string
  is_all_day?: boolean
  timezone?: string
  location?: string
  meeting_url?: string
  meeting_id?: string
  meeting_password?: string
  recurrence_type?: RecurrenceType
  recurrence_pattern?: RecurrencePattern
  recurrence_end_date?: string
  is_private?: boolean
  is_important?: boolean
  reminder_minutes?: number
  color?: string
  meta_data?: Record<string, any>
}

export interface EventListResponse {
  events: CalendarEvent[]
  total: number
  page: number
  page_size: number
  has_next: boolean
}

export interface EventQueryParams {
  start_date?: string
  end_date?: string
  event_type?: EventType
  status?: EventStatus
  project_id?: string
  task_id?: string
  okr_id?: string
  search?: string
  page?: number
  page_size?: number
}

// ============================================================================
// Legacy types for backward compatibility
// ============================================================================

export interface OKREvent {
  id: string
  title: string
  description?: string
  status: string
  timeframe: string
  owner_id: string
  target_date?: string
  progress: number
  created_at: string
  updated_at: string
}

export interface Meeting {
  id: string
  title: string
  description?: string
  start_time: string
  end_time: string
  location?: string
  attendees: string[]
  organizer_id: string
  status: 'scheduled' | 'in_progress' | 'completed' | 'cancelled'
  created_at: string
}

// ============================================================================
// Calendar API Client
// ============================================================================

export class CalendarAPI {
  /**
   * Get calendar events with optional filtering
   */
  static async getEvents(params?: EventQueryParams): Promise<CalendarEvent[]> {
    try {
      const queryParams = new URLSearchParams()

      if (params?.start_date) queryParams.append('start_date', params.start_date)
      if (params?.end_date) queryParams.append('end_date', params.end_date)
      if (params?.event_type) queryParams.append('event_type', params.event_type)
      if (params?.status) queryParams.append('status', params.status)
      if (params?.project_id) queryParams.append('project_id', params.project_id)
      if (params?.task_id) queryParams.append('task_id', params.task_id)
      if (params?.okr_id) queryParams.append('okr_id', params.okr_id)
      if (params?.search) queryParams.append('search', params.search)
      if (params?.page) queryParams.append('page', params.page.toString())
      if (params?.page_size) queryParams.append('page_size', params.page_size.toString())

      const url = `/calendar/events/${queryParams.toString() ? '?' + queryParams.toString() : ''}`
      const response = await api.get<EventListResponse>(url)

      return response.data.events || []
    } catch (error) {
      console.error('Failed to fetch calendar events:', error)
      // Fallback to legacy behavior if calendar API fails
      return this.getLegacyEvents(params?.start_date, params?.end_date)
    }
  }

  /**
   * Get a single event by ID
   */
  static async getEvent(eventId: string): Promise<CalendarEvent> {
    const response = await api.get<CalendarEvent>(`/calendar/events/${eventId}`)
    return response.data
  }

  /**
   * Create a new calendar event
   */
  static async createEvent(eventData: CreateEventRequest): Promise<CalendarEvent> {
    const response = await api.post<CalendarEvent>('/calendar/events/', eventData)
    return response.data
  }

  /**
   * Update an existing event
   */
  static async updateEvent(eventId: string, updates: UpdateEventRequest): Promise<CalendarEvent> {
    const response = await api.patch<CalendarEvent>(`/calendar/events/${eventId}`, updates)
    return response.data
  }

  /**
   * Delete an event
   */
  static async deleteEvent(eventId: string, hardDelete: boolean = false): Promise<void> {
    await api.delete(`/calendar/events/${eventId}?hard_delete=${hardDelete}`)
  }

  /**
   * Get event attendees
   */
  static async getEventAttendees(eventId: string): Promise<EventAttendee[]> {
    const response = await api.get<EventAttendee[]>(`/calendar/events/${eventId}/attendees`)
    return response.data
  }

  /**
   * Update attendee RSVP status
   */
  static async updateAttendeeStatus(
    eventId: string,
    attendeeId: string,
    status: AttendeeStatus,
    notes?: string
  ): Promise<EventAttendee> {
    const response = await api.patch<EventAttendee>(
      `/calendar/events/${eventId}/attendees/${attendeeId}`,
      { status, notes }
    )
    return response.data
  }

  /**
   * Bulk update event status
   */
  static async bulkUpdateStatus(eventIds: string[], status: EventStatus): Promise<number> {
    const response = await api.post<{ affected_count: number }>(
      '/calendar/events/bulk/status',
      { event_ids: eventIds, status }
    )
    return response.data.affected_count
  }

  /**
   * Bulk delete events
   */
  static async bulkDelete(eventIds: string[], deleteRecurring: boolean = false): Promise<number> {
    const response = await api.post<{ affected_count: number }>(
      '/calendar/events/bulk/delete',
      { event_ids: eventIds, delete_recurring: deleteRecurring }
    )
    return response.data.affected_count
  }

  // ============================================================================
  // Helper Methods
  // ============================================================================

  /**
   * Get calendar statistics
   */
  static async getCalendarStats(): Promise<{
    total_events: number
    completed_events: number
    upcoming_events: number
    overdue_events: number
  }> {
    const events = await this.getEvents()
    const now = new Date()

    return {
      total_events: events.length,
      completed_events: events.filter(e => e.status === EventStatus.COMPLETED).length,
      upcoming_events: events.filter(e => new Date(e.start_date) > now).length,
      overdue_events: events.filter(e =>
        new Date(e.start_date) < now &&
        e.status !== EventStatus.COMPLETED &&
        e.status !== EventStatus.CANCELLED
      ).length
    }
  }

  /**
   * Get upcoming deadlines
   */
  static async getUpcomingDeadlines(days: number = 7): Promise<CalendarEvent[]> {
    const now = new Date()
    const futureDate = new Date()
    futureDate.setDate(now.getDate() + days)

    const events = await this.getEvents({
      start_date: now.toISOString(),
      end_date: futureDate.toISOString(),
      page_size: 100
    })

    return events
      .filter(event =>
        event.event_type === EventType.TASK_DEADLINE ||
        event.event_type === EventType.PROJECT_MILESTONE
      )
      .sort((a, b) => new Date(a.start_date).getTime() - new Date(b.start_date).getTime())
  }

  /**
   * Get overdue events
   */
  static async getOverdueEvents(): Promise<CalendarEvent[]> {
    const now = new Date()
    const events = await this.getEvents({ page_size: 100 })

    return events
      .filter(event =>
        new Date(event.start_date) < now &&
        event.status !== EventStatus.COMPLETED &&
        event.status !== EventStatus.CANCELLED
      )
      .sort((a, b) => new Date(a.start_date).getTime() - new Date(b.start_date).getTime())
  }

  /**
   * Convert event to legacy format for compatibility
   */
  static toLegacyFormat(event: CalendarEvent): any {
    return {
      id: event.id,
      title: event.title,
      start: new Date(event.start_date),
      end: new Date(event.end_date),
      type: event.event_type,
      description: event.description,
      status: event.status,
      tags: [event.event_type, event.status],
      allDay: event.is_all_day,
      recurring: event.recurrence_type !== RecurrenceType.NONE,
      location: event.location,
      attendees: event.attendees?.map(a => a.display_name || a.email || ''),
      project_id: event.project_id,
      task_id: event.task_id,
      okr_id: event.okr_id
    }
  }

  // ============================================================================
  // Legacy Methods (fallback to tasks/OKRs/projects)
  // ============================================================================

  /**
   * Legacy method: Get events from tasks, OKRs, and projects
   * Used as fallback if calendar API is unavailable
   */
  private static async getLegacyEvents(startDate?: string, endDate?: string): Promise<CalendarEvent[]> {
    const events: CalendarEvent[] = []

    // Fetch tasks
    try {
      const tasksResponse = await api.get('/tasks/')
      const tasks: Task[] = tasksResponse.data

      const taskEvents: CalendarEvent[] = tasks
        .filter(task => task.due_date)
        .map(task => this.taskToCalendarEvent(task))

      events.push(...taskEvents)
    } catch (error) {
      console.error('Failed to fetch tasks for calendar:', error)
    }

    // Fetch OKRs
    try {
      const okrsResponse = await api.get('/okrs')
      const okrs: OKREvent[] = okrsResponse.data

      const okrEvents: CalendarEvent[] = okrs
        .filter(okr => okr.target_date)
        .map(okr => this.okrToCalendarEvent(okr))

      events.push(...okrEvents)
    } catch (error) {
      console.error('Failed to fetch OKRs for calendar:', error)
    }

    // Fetch projects
    try {
      const projectsResponse = await api.get('/projects')
      const projects = projectsResponse.data

      const milestoneEvents: CalendarEvent[] = projects
        .filter((project: any) => project.end_date)
        .map((project: any) => this.projectToCalendarEvent(project))

      events.push(...milestoneEvents)
    } catch (error) {
      console.error('Failed to fetch projects for calendar:', error)
    }

    // Filter by date range if provided
    if (startDate && endDate) {
      const start = new Date(startDate)
      const end = new Date(endDate)
      return events.filter(event => {
        const eventStart = new Date(event.start_date)
        return eventStart >= start && eventStart <= end
      })
    }

    return events
  }

  private static taskToCalendarEvent(task: Task): CalendarEvent {
    return {
      id: `task-${task.id}`,
      title: task.title,
      description: task.description,
      event_type: EventType.TASK_DEADLINE,
      status: task.status === TaskStatus.DONE ? EventStatus.COMPLETED : EventStatus.CONFIRMED,
      start_date: task.due_date!,
      end_date: task.due_date!,
      is_all_day: true,
      timezone: 'UTC',
      recurrence_type: RecurrenceType.NONE,
      is_private: false,
      is_important: task.priority === TaskPriority.CRITICAL || task.priority === TaskPriority.HIGH,
      color: this.getPriorityColor(task.priority),
      org_id: '',
      created_at: task.created_at,
      updated_at: task.updated_at,
      is_deleted: false,
      task_id: task.id,
      project_id: task.project_id
    }
  }

  private static okrToCalendarEvent(okr: OKREvent): CalendarEvent {
    return {
      id: `okr-${okr.id}`,
      title: `🌟 ${okr.title}`,
      description: okr.description,
      event_type: EventType.OKR_REVIEW,
      status: EventStatus.CONFIRMED,
      start_date: okr.target_date!,
      end_date: okr.target_date!,
      is_all_day: true,
      timezone: 'UTC',
      recurrence_type: RecurrenceType.NONE,
      is_private: false,
      is_important: true,
      color: '#f59e0b',
      org_id: '',
      created_at: okr.created_at,
      updated_at: okr.updated_at,
      is_deleted: false,
      okr_id: okr.id
    }
  }

  private static projectToCalendarEvent(project: any): CalendarEvent {
    return {
      id: `project-${project.id}`,
      title: `📄 ${project.name} Deadline`,
      description: project.description,
      event_type: EventType.PROJECT_MILESTONE,
      status: EventStatus.CONFIRMED,
      start_date: project.end_date,
      end_date: project.end_date,
      is_all_day: true,
      timezone: 'UTC',
      recurrence_type: RecurrenceType.NONE,
      is_private: false,
      is_important: true,
      color: '#8b5cf6',
      org_id: '',
      created_at: project.created_at || new Date().toISOString(),
      updated_at: project.updated_at || new Date().toISOString(),
      is_deleted: false,
      project_id: project.id
    }
  }

  private static getPriorityColor(priority?: TaskPriority): string {
    switch (priority) {
      case TaskPriority.CRITICAL:
        return '#ef4444'
      case TaskPriority.HIGH:
        return '#f97316'
      case TaskPriority.MEDIUM:
        return '#3b82f6'
      case TaskPriority.LOW:
        return '#10b981'
      default:
        return '#6b7280'
    }
  }
}

File: apps/api/liderix_api/schemas/calendar.py
"""
Calendar and Event Pydantic schemas for API request/response validation
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from liderix_api.enums import EventType, EventStatus, RecurrenceType


# ============================================================================
# Event Attendee Schemas
# ============================================================================

class EventAttendeeBase(BaseModel):
    """Base schema for event attendees"""
    user_id: Optional[UUID] = Field(None, description="User ID (for internal users)")
    email: Optional[str] = Field(None, description="Email (for external attendees)")
    display_name: Optional[str] = Field(None, description="Display name (for external attendees)")
    is_required: bool = Field(True, description="Is attendance required")
    notes: Optional[str] = Field(None, description="Attendee notes")


class EventAttendeeCreate(EventAttendeeBase):
    """Schema for creating event attendee"""
    pass


class EventAttendeeUpdate(BaseModel):
    """Schema for updating attendee status"""
    status: str = Field(..., description="pending, accepted, declined, tentative")
    notes: Optional[str] = Field(None, description="Response notes")


class EventAttendeeResponse(EventAttendeeBase):
    """Schema for event attendee response"""
    id: UUID
    event_id: UUID
    status: str
    is_organizer: bool
    response_date: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============================================================================
# Recurrence Pattern Schema
# ============================================================================

class RecurrencePattern(BaseModel):
    """Recurrence configuration schema"""
    interval: int = Field(1, ge=1, description="Repeat every N units (e.g., every 2 weeks)")
    days_of_week: Optional[List[int]] = Field(
        None,
        description="Days of week (0=Monday, 6=Sunday) for weekly recurrence"
    )
    day_of_month: Optional[int] = Field(
        None,
        ge=1,
        le=31,
        description="Day of month for monthly recurrence"
    )
    month_of_year: Optional[int] = Field(
        None,
        ge=1,
        le=12,
        description="Month of year for yearly recurrence"
    )
    count: Optional[int] = Field(
        None,
        ge=1,
        description="Number of occurrences (alternative to end_date)"
    )
    exceptions: Optional[List[datetime]] = Field(
        None,
        description="Exception dates (EXDATE in RRULE)"
    )


# ============================================================================
# Calendar Event Schemas
# ============================================================================

class CalendarEventBase(BaseModel):
    """Base schema for calendar events"""
    title: str = Field(..., min_length=1, max_length=500, description="Event title")
    description: Optional[str] = Field(None, description="Event description")
    event_type: EventType = Field(EventType.MEETING, description="Event type")
    start_date: datetime = Field(..., description="Event start date/time (ISO 8601)")
    end_date: datetime = Field(..., description="Event end date/time (ISO 8601)")
    is_all_day: bool = Field(False, description="Is this an all-day event")
    timezone: str = Field("UTC", description="Event timezone (IANA format)")

    # Location and meeting
    location: Optional[str] = Field(None, max_length=500, description="Physical location")
    meeting_url: Optional[str] = Field(None, max_length=1000, description="Video meeting URL")
    meeting_id: Optional[str] = Field(None, max_length=100, description="Meeting ID")
    meeting_password: Optional[str] = Field(None, max_length=100, description="Meeting password")

    # Recurrence
    recurrence_type: RecurrenceType = Field(
        RecurrenceType.NONE,
        description="Recurrence type"
    )
    recurrence_pattern: Optional[RecurrencePattern] = Field(
        None,
        description="Recurrence configuration"
    )
    recurrence_end_date: Optional[datetime] = Field(
        None,
        description="End date for recurring events"
    )

    # Settings
    is_private: bool = Field(False, description="Is event private")
    is_important: bool = Field(False, description="Is event important/high priority")
    reminder_minutes: Optional[int] = Field(None, ge=0, description="Reminder before event (minutes)")
    color: Optional[str] = Field("#3174ad", pattern="^#[0-9A-Fa-f]{6}$", description="Hex color")

    # Metadata
    meta_data: Optional[Dict[str, Any]] = Field(None, description="Additional metadata")

    @model_validator(mode='after')
    def validate_dates(self) -> 'CalendarEventBase':
        """Validate that end_date is after start_date"""
        if self.end_date <= self.start_date:
            raise ValueError("end_date must be after start_date")
        return self

    @field_validator('recurrence_pattern')
    def validate_recurrence(cls, v: Optional[RecurrencePattern], info) -> Optional[RecurrencePattern]:
        """Validate recurrence pattern is provided when recurrence_type is not NONE"""
        if info.data.get('recurrence_type') != RecurrenceType.NONE and not v:
            raise ValueError("recurrence_pattern is required when recurrence_type is not NONE")
        return v


class CalendarEventCreate(CalendarEventBase):
    """Schema for creating calendar events"""
    # Related entities (optional)
    task_id: Optional[UUID] = Field(None, description="Related task ID")
    project_id: Optional[UUID] = Field(None, description="Related project ID")
    okr_id: Optional[UUID] = Field(None, description="Related OKR ID")

    # Attendees
    attendees: Optional[List[EventAttendeeCreate]] = Field(
        None,
        description="Event attendees"
    )


class CalendarEventUpdate(BaseModel):
    """Schema for updating calendar events (all fields optional)"""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    event_type: Optional[EventType] = None
    status: Optional[EventStatus] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_all_day: Optional[bool] = None
    timezone: Optional[str] = None
    location: Optional[str] = None
    meeting_url: Optional[str] = None
    meeting_id: Optional[str] = None
    meeting_password: Optional[str] = None
    recurrence_type: Optional[RecurrenceType] = None
    recurrence_pattern: Optional[RecurrencePattern] = None
    recurrence_end_date: Optional[datetime] = None
    is_private: Optional[bool] = None
    is_important: Optional[bool] = None
    reminder_minutes: Optional[int] = Field(None, ge=0)
    color: Optional[str] = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    meta_data: Optional[Dict[str, Any]] = None

    @model_validator(mode='after')
    def validate_dates_if_both_present(self) -> 'CalendarEventUpdate':
        """Validate dates only if both are provided"""
        if self.start_date and self.end_date and self.end_date <= self.start_date:
            raise ValueError("end_date must be after start_date")
        return self


class CalendarEventResponse(CalendarEventBase):
    """Schema for calendar event response"""
    id: UUID
    org_id: UUID
    creator_id: Optional[UUID] = None
    status: EventStatus

    # Related entities
    task_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    okr_id: Optional[UUID] = None

    # Recurrence
    parent_event_id: Optional[UUID] = None

    # External
    external_id: Optional[str] = None
    external_source: Optional[str] = None

    # Timestamps
    created_at: datetime
    updated_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    is_deleted: bool = False

    # Relationships
    attendees: List[EventAttendeeResponse] = Field(default_factory=list)

    # Computed
    duration_minutes: Optional[int] = Field(None, description="Event duration in minutes")

    class Config:
        from_attributes = True


class CalendarEventListResponse(BaseModel):
    """Schema for paginated event list response"""
    events: List[CalendarEventResponse]
    total: int
    page: int = 1
    page_size: int = 50
    has_next: bool = False


# ============================================================================
# Calendar Container Schemas
# ============================================================================

class CalendarBase(BaseModel):
    """Base schema for calendar containers"""
    name: str = Field(..., min_length=1, max_length=200, description="Calendar name")
    description: Optional[str] = Field(None, description="Calendar description")
    color: str = Field("#3174ad", pattern="^#[0-9A-Fa-f]{6}$", description="Calendar color")
    is_public: bool = Field(False, description="Is calendar publicly visible")
    timezone: str = Field("UTC", description="Calendar timezone")
    default_reminder_minutes: Optional[int] = Field(15, ge=0, description="Default reminder time")


class CalendarCreate(CalendarBase):
    """Schema for creating calendars"""
    is_default: bool = Field(False, description="Set as default calendar for user")


class CalendarUpdate(BaseModel):
    """Schema for updating calendars (all fields optional)"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    color: Optional[str] = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    is_public: Optional[bool] = None
    is_active: Optional[bool] = None
    timezone: Optional[str] = None
    default_reminder_minutes: Optional[int] = Field(None, ge=0)


class CalendarResponse(CalendarBase):
    """Schema for calendar response"""
    id: UUID
    org_id: UUID
    owner_id: UUID
    is_default: bool
    is_active: bool
    sync_enabled: bool
    last_sync_at: Optional[datetime] = None
    external_id: Optional[str] = None
    external_source: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CalendarListResponse(BaseModel):
    """Schema for calendar list response"""
    calendars: List[CalendarResponse]
    total: int


# ============================================================================
# Calendar Permission Schemas
# ============================================================================

class CalendarPermissionBase(BaseModel):
    """Base schema for calendar permissions"""
    user_id: UUID = Field(..., description="User to grant permission to")
    permission: str = Field(..., description="Permission level: read, write, admin")

    @field_validator('permission')
    def validate_permission(cls, v: str) -> str:
        """Validate permission level"""
        if v not in ['read', 'write', 'admin']:
            raise ValueError("permission must be one of: read, write, admin")
        return v


class CalendarPermissionCreate(CalendarPermissionBase):
    """Schema for creating calendar permission"""
    pass


class CalendarPermissionUpdate(BaseModel):
    """Schema for updating calendar permission"""
    permission: str = Field(..., description="Permission level: read, write, admin")

    @field_validator('permission')
    def validate_permission(cls, v: str) -> str:
        """Validate permission level"""
        if v not in ['read', 'write', 'admin']:
            raise ValueError("permission must be one of: read, write, admin")
        return v


class CalendarPermissionResponse(CalendarPermissionBase):
    """Schema for calendar permission response"""
    id: UUID
    calendar_id: UUID
    granted_by_id: Optional[UUID] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============================================================================
# Query Parameter Schemas
# ============================================================================

class EventQueryParams(BaseModel):
    """Schema for event query parameters"""
    start_date: Optional[datetime] = Field(None, description="Filter events starting from this date")
    end_date: Optional[datetime] = Field(None, description="Filter events until this date")
    event_type: Optional[EventType] = Field(None, description="Filter by event type")
    status: Optional[EventStatus] = Field(None, description="Filter by status")
    task_id: Optional[UUID] = Field(None, description="Filter by related task")
    project_id: Optional[UUID] = Field(None, description="Filter by related project")
    okr_id: Optional[UUID] = Field(None, description="Filter by related OKR")
    creator_id: Optional[UUID] = Field(None, description="Filter by creator")
    is_private: Optional[bool] = Field(None, description="Filter by privacy")
    search: Optional[str] = Field(None, description="Search in title and description")
    page: int = Field(1, ge=1, description="Page number")
    page_size: int = Field(50, ge=1, le=100, description="Items per page")
    sort_by: str = Field("start_date", description="Sort field")
    sort_order: str = Field("asc", description="Sort order: asc or desc")

    @field_validator('sort_order')
    def validate_sort_order(cls, v: str) -> str:
        """Validate sort order"""
        if v not in ['asc', 'desc']:
            raise ValueError("sort_order must be 'asc' or 'desc'")
        return v


# ============================================================================
# Bulk Operations
# ============================================================================

class BulkEventStatusUpdate(BaseModel):
    """Schema for bulk event status update"""
    event_ids: List[UUID] = Field(..., min_length=1, max_length=100, description="Event IDs to update")
    status: EventStatus = Field(..., description="New status")


class BulkEventDelete(BaseModel):
    """Schema for bulk event deletion"""
    event_ids: List[UUID] = Field(..., min_length=1, max_length=100, description="Event IDs to delete")
    delete_recurring: bool = Field(False, description="Delete all recurring instances")


# ============================================================================
# Response Messages
# ============================================================================

class EventOperationResponse(BaseModel):
    """Generic response for event operations"""
    message: str
    event_id: Optional[UUID] = None
    affected_count: Optional[int] = None
